src/storage/src/render/upsert.rs

```diff
@@ -312,18 +246,37 @@ where
     builder.build(move |caps| async move {
         let mut output_cap = caps.into_element();
 
-        let mut snapshot = vec![];
+        let mut state = StatsState::new(state().await, upsert_shared_metrics);
 
-        // In the first phase we will collect all the updates from our output that are not beyond
-        // resume_upper. This will be the seed state for the command processing below.
+        let mut batch_key_counter = HashSet::with_capacity(US::SNAPSHOT_BATCH_SIZE);
+        let mut update_buf = Vec::with_capacity(US::SNAPSHOT_BATCH_SIZE * 10);
+        let now = Instant::now();
+        let mut stats = types::MergeStats::default();
         while let Some(event) = previous.next_mut().await {
             match event {
                 AsyncEvent::Data(_cap, data) => {
-                    snapshot.extend(
-                        data.drain(..)
-                            .filter(|(_row, ts, _diff)| !resume_upper.less_equal(ts))
-                            .map(|(row, _ts, diff)| (row, diff)),
-                    );
+                    for ((key, value), ts, diff) in data.drain(..) {
+                        // In the first phase we consolidate updates from our output that are not beyond
+                        // resume_upper, in-place in `state`.
+                        #[allow(clippy::as_conversions)]
+                        if !resume_upper.less_equal(&ts) {
+                            update_buf.push((key, value, diff));
+                            batch_key_counter.insert(key);
+
+                            // If the number of keys in this batch is >= the requested
+                            // size, flush it out. We also flush if there are >= `SNAPSHOT_BATCH_SIZE`*10
+                            // updates, with < SNAPSHOT_BATCH_SIZE keys.
+                            if batch_key_counter.len() >= US::SNAPSHOT_BATCH_SIZE
+                                || update_buf.len() >= US::SNAPSHOT_BATCH_SIZE * 10
```

Comment:
I'm not sure if this is worth doing. When an operator runs in timely there is a certain amount of data ready to go in a handle, stored in memory. What we want to do is drain all the data in the input handle and process it in the upsert store, which you can do by calling `now_or_never()` on the input handle. With the fixed batch size you're risking accumulating work in memory that could be retired without touching the hashmap at all.

Ideally the `AsyncInputHandle` should have a `.ready()` and `next_sync()` API to enable this pattern without resorting to `now_or_never()` but I suppose you don't want to do this in this PR (but if you do I fully support it!)

Reply:
I just tested this out locally and using `.now_or_never()` is pretty natural. I think this section should be with the XOR logic in here instead of wrapping the `UpsertStateBackend` in `UpsertState` because this logic is completely independent of what we use to store the data in. 

```rust
        let mut events = vec![];
        let mut snapshot_upper = Antichain::from_elem(Timestamp::minimum());
        while !PartialOrder::less_equal(&resume_upper, &snapshot_upper) {
            match previous.next_mut() {
                Some(AsyncEvent::Data(_cap, data)) => events.extend(data.drain(..)),
                Some(AsyncEvent::Progress(upper)) => snapshot_upper = upper,
                None => snapshot_upper = Antichain::new(),
            };
            while let Some(event) = previous.next_mut().now_or_never() {
                match event {
                    Some(AsyncEvent::Data(_cap, data)) => events.extend(data.drain(..)),
                    Some(AsyncEvent::Progress(upper)) => snapshot_upper = upper,
                    None => snapshot_upper = Antichain::new(),
                }
            }
            // 1. consolidate events in memory
            // 2. multi_get the values of all keys in `events`
            // 3. perform in-memory XOR
            // 4. multi_put the final in-memory state
        }
```